### PR TITLE
Virtual null values

### PR DESCRIPTION
**Pull request by emrysr**
Check query response when describing table in db_schema_update_column().
Returned NULL value for get_timevalue() virtual feeds with no lastvalue
added comments
Hidden E_DEPRECATED PHP warnings as they'll break any json responses

delwedd

### CODE REVIEW

**Comment by TrystanLea** [comment]
@chaveiro could you review this proposed change to virtual feeds?

File: Modules/feed/feed_model.php [hunk 1]
```diff
@@ -516,7 +516,12 @@ public function get_timevalue($id)
         if ($engine == Engine::VIRTUALFEED) { //if virtual get it now
             $this->log->info("get_timevalue() calling VIRTUAL lastvalue $id");
             $lastvirtual = $this->EngineClass(Engine::VIRTUALFEED)->lastvalue($id);
-            return array('time'=>$lastvirtual['time'], 'value'=>$lastvirtual['value']);
+            if ($lastvirtual) {
+                $timevalue = array('time'=>$lastvirtual['time'], 'value'=>$lastvirtual['value']);
+            } else {
+                $timevalue = array('time'=>null, 'value'=>null);
```
Comment (chaveiro) [inline Modules/feed/feed_model.php]: What problem is this fixing?

File: Lib/dbschemasetup.php [hunk 2]
```diff
@@ -172,13 +177,21 @@ function db_schema_add_column($mysqli, $table, $field, $spec, &$operations)
...
-    $result = $mysqli->query("DESCRIBE $table `$field`");
+    $result = $mysqli->query("DESCRIBE `$table` `$field`");
     $current = $result->fetch_array();
 
     // Check if we 1. need to change type 2. need to add a primary key
     $diff_datatype = db_schema_diff_datatype($spec, $current, $field);
     $add_key = db_schema_diff_key($spec, $current);
 
+
+    // if($table === 'feeds' && $field === 'id') {
```
Comment (chaveiro) [inline Lib/dbschemasetup.php]: Please try to keep code clean

File: process_settings.php [hunk 3]
```diff
@@ -63,8 +63,10 @@
 
 // Set display errors
 if (isset($settings["display_errors"]) && ($settings["display_errors"])) {
-    error_reporting(E_ALL);
+    // @todo: add setting to allow for deprecated warnings
     ini_set('display_errors', 'on');
+    // show all errors except for E_DEPRECATED warnings
+    error_reporting(E_ALL & ~E_DEPRECATED);
```
Comment (chaveiro) [inline process_settings.php]: Why omit the error if we could fix the code?

File: Modules/feed/feed_model.php (hunk 1 above)
Comment (emrysr) [reply Modules/feed/feed_model.php]: has hit errors when $lastvirtual is null and has no ['time'] or ['value']. so I checked for that and returned null values for the `time` and `value`.

File: process_settings.php (hunk 3 above)
Comment (emrysr) [reply process_settings.php]: The deprecated messages are warnings (nothing wrong with the code), however they'll print out and break any API responses that require a specific output format.

Comment (emrysr) [reply process_settings.php]: If you'd like to keep all errors and warning messages showing I'll change it back.

File: Modules/feed/feed_model.php (hunk 1 above)
Comment (chaveiro) [reply Modules/feed/feed_model.php]: **Time** must not be null but **value** can be null.
This fix should go to where the original error came from, the engine class lastvalue() function.

File: process_settings.php (hunk 3 above)
Comment (chaveiro) [reply process_settings.php]: They are warning that on a next version of PHP the code will not work as the functions are deprecated.
Why wait until it brokes if you can fix it now?

**Comment by chaveiro** [comment]
Comments added to the previous review.

**Comment by emrysr** [comment]
this pull request should be ok now. all the confusion taken out.

**Comment by TrystanLea** [comment]
This issue was a bit of a strange issue, we observed two different behaviours on two different systems, if the virtual feed last value function returned false on one system, the feed_model would ignore the missing time and value keys and populate them as null anyway, on the other system this same section would throw the expected undefined error...

**Pull request merged**